namer/rancher/src/main/scala/io/buoyant/namer/rancher/RancherClient.scala

```diff
@@ -0,0 +1,149 @@
+import com.twitter.logging.Logger
+import com.twitter.util.Activity.State
+import com.twitter.util.{Timer, Var, Activity, Duration, Future, Closable, Return, Throw}
+import java.util.concurrent.atomic.AtomicBoolean
+import scala.reflect.{ClassTag, classTag}
+
+case class PortMapping(
+  ip: String,
+  publicPort: Int,
+  privatePort: Int,
+  protocol: String
+) {
+  override def toString(): String = s"$ip:$privatePort:$publicPort/$protocol"
+}
+class PortMappingDeserializer extends StdDeserializer[PortMapping](RancherParser.jClass[PortMapping]) {
+  private[this] val Mapping = raw"([\.0-9]+):([0-9]+):([0-9]+)/(tcp|udp)".r
+  override def deserialize(jp: JsonParser, ctx: DeserializationContext): PortMapping = {
+    val mapping = _parseString(jp, ctx)
+    mapping match {
+      case Mapping(ip, privatePort, publicPort, protocol) =>
+        PortMapping(ip, privatePort.toInt, publicPort.toInt, protocol)
+      case _ =>
+        throw ctx.mappingException(s"Failed to understand ip-binding: $mapping")
+    }
+  }
+}
+
+@JsonIgnoreProperties(ignoreUnknown = true)
+case class RancherContainer(
+  createIndex: Int,
+  ips: List[String],
+  primaryIp: String,
+  labels: Map[String, String],
+  name: String,
+  ports: List[PortMapping],
+  serviceName: String,
+  stackName: String,
+  state: String
+) {
+  def toAddr(port: Int): Option[Address] = state match {
+    case "running" => Some(Address(primaryIp, port))
+    case _ => None
+  }
+}
+
+object RancherParser {
+  private[rancher] def jClass[T: ClassTag] = classTag[T].runtimeClass.asInstanceOf[Class[T]]
+
+  private[this] val customTypes = new SimpleModule("RancherClient custom types")
+  customTypes.addDeserializer(
+    RancherParser.jClass[PortMapping],
+    new PortMappingDeserializer()
+  )
+  private[this] val objectMapper = new ObjectMapper() with ScalaObjectMapper
+  objectMapper.registerModule(DefaultScalaModule)
+  objectMapper.registerModule(customTypes)
+  objectMapper.setPropertyNamingStrategy(PropertyNamingStrategy.SNAKE_CASE)
+
+  def apply() = objectMapper
+}
+
+class RancherClient(
+  maxWait: Int,
+  log: Logger,
+  params: Stack.Params,
+  stats: StatsReceiver
+)(implicit val timer: Timer) {
+  private[this] val numVersions = stats.counter("metadata_versions")
+
+  private[this] val service = Http.client
+    .withStack(Http.client.stack.remove(Retries.Role))
+    .withParams(Http.client.params ++ params)
+    .withLabel("io.buoyant.namer.rancher.client")
+    .withTracer(NullTracer)
+    .newService(s"/$$/inet/rancher-metadata/80")
+
+  private[this] val version: Activity[String] =
+    Activity(Var.async[State[String]](Activity.Pending) { state =>
+      val done = new AtomicBoolean(false)
+      @volatile var v: Option[String] = None
+
+      Future.whileDo(!done.get) {
+        log.trace("starting version request")
+
+        val request = http.Request(http.Method.Get, s"/2015-12-19/version?wait=true&maxWait=${maxWait}&value=${v}")
```

Comment:
Just leaving this here until I have the time to fix it: `${v}` needs to be updated after `v` changed from `String` to a `Option[String]` - so probably something like `${v.getOrElse('initial')}`.